apps/aechannel/src/aesc_fsm_min_depth_watcher.erl

```diff
@@ -104,122 +239,707 @@ terminate(_Reason, _St) ->
+    H >= H1;
+check_req_at_height(_, _) ->
+    false.
+
+%% May involve chain access
+check_requests(Reqs, St, C) ->
+    do_dirty(
+      fun() ->
+              check_requests(Reqs, St, C, [])
+      end).
+
+%% Trap exceptions INSIDE the db activity for meaningful error reporting
+check_requests(Reqs, St, C, []) ->
+    Reqs1 = reset_if_fork_switch(Reqs, C),
+    try check_requests_(Reqs1, St, C, [])
+    catch
+        error:E ->
+            lager:error("CAUGHT ~p / ~p", [E, erlang:get_stacktrace()]),
+            error(E)
+    end.
+
+%% If there's a fork switch, don't assume channel state or depth calculations
+reset_if_fork_switch(Reqs, #{scenario := S}) ->
+    case S of
+        fork_switch ->
+            [maps:remove(check_at_height,
+                         maps:remove(locked_until, R)) || R <- Reqs];
         _ ->
-            {R, Cache1}
+            Reqs
+    end.
+
+init_cache(top, St) ->
+    C0 = init_cache_(St),
+    %% Set only top_hash. Further top_info fetched on-demand
+    {_THash, C1} = top_hash(C0),
+    C1;
+init_cache(#{ block_hash := BHash
+            , prev_hash  := PHash
+            , block_type := Type
+            , height     := Height }, St) ->
+    C0 = init_cache_(St),
+    %% Here, we already have all the top_info, so cache it
+    C0#{ top_hash => BHash
+       , top_info => #{ block_hash => BHash
+                      , prev_hash  => PHash
+                      , block_type => Type
+                      , height     => Height } }.
+
+init_cache_(#st{ last_block = LastBlock
+              , last_top   = LastTop
+              , tx_log     = TxLog
+              , rpt_log    = RptLog
+              , chan_vsn   = Vsn }) ->
+    #{ tx_log     => TxLog
+     , rpt_log    => RptLog
+     , last_block => LastBlock  %% Last block actually checked
+     , last_top   => LastTop    %% Last top event hash (possibly skipped)
+     , chan_vsn   => Vsn }.
+
+%% The TxLog relies on the custom tx events (remember that it's a bounded list)
+find_tx_in_block(BHash, TxLog) ->
+    case aesc_window:info_find([{block_hash, BHash}], 2, TxLog) of
+        false ->
+            false;
+        LastTx ->
+            LastTx
+    end.
+
+%% We do an async_dirty activity for minimal overhead. Note that the analysis
+%% is already fixed by the top_hash, so it should be deterministic even in
+%% dirty mode. Db accesses are cached so they only need to happen once.
+do_dirty(F) ->
+    aec_db:ensure_activity(
+      async_dirty,
+      fun() ->
+              try F()
+              catch
+                  error:E ->
+                      lager:error("CAUGHT ~p / ~p", [E, erlang:get_stacktrace()]),
+                      error(E)
+              end
+      end).
+
+-spec check_requests_([req()], #st{}, cache(), [req()]) -> #st{}.
+check_requests_([#{check_at_height := CheckAt} = R|Reqs], St, C, Acc) ->
+    lager:debug("CheckAt: R = ~p", [R]),
+    {TopHeight, C1} = top_height(C),
```

Comment:
Is new cache `C1` here unused on purpose?

Reply:
No, it's a bug.

Reply:
I see addressed.